### PR TITLE
Maintenance: automate header guards

### PR DESCRIPTION
**Pull request by kinkie**
Add a source-maintenance plugin to maintain a
header guard standard defined as SQUID_FILE_PATH_H.

### CODE REVIEW

**Comment by kinkie** [comment]
Please advise whether the PR should include the changes from the execution or to leave that as the subsequent run of source-maintenance

**Comment by rousskov** [comment]
> Please advise whether the PR should include the changes from the execution

IMO, it should not because keeping large automated charges outside decreases this PR noise and helps with porting. However, I would not object to the inclusion if others insist on it. It is not a big deal either way.

Thank you for posting the diff!

**Comment by kinkie** [comment]
Small change needed in src/dns/rfc3596.cc as it relies on header guard internals, breaking layering

File: src/dns/rfc3596.cc
Comment (yadij) [inline src/dns/rfc3596.cc:28]: Since you are changing this manually, please update to match our current requirements for non-file wrappers:
```suggestion
#if !defined(RFC1035_MAXHOSTNAMESZ)
```

File: scripts/maintenance/standard-header-guards.pl
Comment (yadij) [inline scripts/maintenance/standard-header-guards.pl:64]: Missing empty line at the end of the file. This is needed to avoid issues with some of our tools.
```suggestion
print "#endif /* $new_guard */\n\n";

```

File: scripts/maintenance/standard-header-guards.pl
Comment (yadij) [inline scripts/maintenance/standard-header-guards.pl:41]: Please consistently use either `exit(0)` or `exit 0`. Currently there is a mix.

**Review by yadij — CHANGES_REQUESTED** [review]

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:41]: Good catch; changing to also use function-call syntax for calls to die()

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:41]: Please quit with an error if a `.h` file in `src/` does not have a guard (that the script can detect). I counted approximately 15 such files (that all should be fixed AFAICT):

```
$ git grep -L -P '^#ifndef\s+(\w+_H)' '*.h' | fgrep src/
src/acl/AnnotateClient.h
src/acl/AnnotateTransaction.h
src/acl/external/kerberos_ldap_group/support.h
src/adaptation/ecap/Registry.h
src/auth/State.h
src/auth/basic/NIS/nis_support.h
src/auth/basic/RADIUS/radius-util.h
src/auth/basic/RADIUS/radius.h
src/auth/digest/LDAP/ldap_backend.h
src/auth/digest/eDirectory/edir_ldapext.h
src/auth/digest/eDirectory/ldap_backend.h
src/auth/digest/file/text_backend.h
src/auth/negotiate/kerberos/negotiate_kerberos.h
src/tests/STUB.h
src/tests/TestSwapDir.h
```

P.S. Please fix the "fine" typo in the comment while you are adjusting this code.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:64]: Please refactor to quit with an error if the expected terminating `#endif` is not found (instead of silently adding a matching `#endif`) and if there are unexpected accumulated leftovers after the expected terminating `#endif` (instead of silently dropping them).

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:46]: Instead of rematching to set current_guard after the guard-searching loop, I recommend setting current_guard inside that loop itself.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:51]: For readability sake:

```suggestion
print "#ifndef $new_guard\n";
print "#define $new_guard\n";
```

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for adding this script.

**Comment by rousskov** [comment]
I removed temporary references from the PR description and removed RELATIVE from the pattern there. IMHO, "relative" is confusing without a reference point (i.e. "relative to what?"). I do not think we have to detail that -- the script and adjusted sources will speak for themselves. I do not insist on removing RELATIVE though.

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:48]: ok

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:66]: The side effect of this change is that it requires including in the PR all the problematic header files, otherwise the source-maintenance Github check will fail this requiring some hoop-jumping to land the code. Now done

**Comment by kinkie** [comment]
@rousskov I think we should be good. I had to hardcode ignoring headers in scripts/ , I couldn't think of any better way to avoid clobbering scripts/boilerplate.h without affecting its function

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [reply scripts/maintenance/standard-header-guards.pl:72]: Yes, and this is a _good_ side effect IMO. If others disagree, we can fix those files in a prerequisite minimal PR instead.

**Comment by rousskov** [comment]
> I had to hardcode ignoring headers in scripts/ , I couldn't think of any better way to avoid clobbering scripts/boilerplate.h without affecting its function

My [suggestion](https://github.com/squid-cache/squid/pull/1630#discussion_r1445233575) was to limit the corresponding fatal errors to headers in `src/`. I will evaluate your alternative approach when I review this PR.

**Comment by kinkie** [comment]
> > I had to hardcode ignoring headers in scripts/ , I couldn't think of any better way to avoid clobbering scripts/boilerplate.h without affecting its function
> 
> My [suggestion](https://github.com/squid-cache/squid/pull/1630#discussion_r1445233575) was to limit the corresponding fatal errors (and, hence, formatting) to headers in `src/`. I will evaluate your alternative approach when I review this PR.

we also have heaaders in include/ and lib/. In practice it seems to work well, the only file that shouldn't have header guards seems to be boilerplate.h. Looking forward to hear if you agree with my approach

File: src/adaptation/ecap/Registry.h
Comment (yadij) [inline src/adaptation/ecap/Registry.h:14]:
```suggestion
#endif /* SQUID_SRC_ADAPTATION_ECAP_REGISTRY_H */

```

File: src/acl/AnnotateTransaction.h
Comment (yadij) [inline src/acl/AnnotateTransaction.h:10]: It would be best to add the new macro syntax to avoid unnecessary double-touching and adjust the `#endif` as well for clarity that this was not broken earlier.
```suggestion
#ifndef SQUID_SRC_ACL_ANNOTATETRANSACTION_H
#define SQUID_SRC_ACL_ANNOTATETRANSACTION_H
```

File: src/auth/basic/NIS/nis_support.h
Comment (yadij) [inline src/auth/basic/NIS/nis_support.h:14]:
```suggestion
#endif /* SQUID_SRC_AUTH_BASIC_NIS_NIS_SUPPORT_H */

```

File: src/auth/basic/RADIUS/radius-util.h
Comment (yadij) [inline src/auth/basic/RADIUS/radius-util.h:20]:
```suggestion
#endif /* SQUID_SRC_AUTH_BASIC_RADIUS_RADIUS_UTIL_H */

```

File: src/auth/digest/LDAP/ldap_backend.h
Comment (yadij) [inline src/auth/digest/LDAP/ldap_backend.h:21]:
```suggestion
#endif /* SQUID_SRC_AUTH_DIGEST_LDAP_LDAP_BACKEND_H */

```

File: src/auth/digest/eDirectory/ldap_backend.h
Comment (yadij) [inline src/auth/digest/eDirectory/ldap_backend.h:20]:
```suggestion
#endif /* SQUID_SRC_AUTH_DIGEST_EDIRECTORY_LDAP_BACKEND_H */

```

File: src/auth/negotiate/kerberos/negotiate_kerberos.h
Comment (yadij) [inline src/auth/negotiate/kerberos/negotiate_kerberos.h:154]:
```suggestion
#endif /* SQUID_SRC_ACL_EXTERNAL_KERBEROS_LDAP_GROUP_SUPPORT_H */

```

File: src/tests/STUB.h
Comment (yadij) [inline src/tests/STUB.h:57]:
```suggestion
#endif /* SQUID_SRC_TESTS_STUB_H */

```

File: src/tests/TestSwapDir.h
Comment (yadij) [inline src/tests/TestSwapDir.h:10]:
```suggestion
#ifndef SQUID_SRC_TESTS_TESTSWAPDIR_H
#define SQUID_SRC_TESTS_TESTSWAPDIR_H
```

File: src/tests/TestSwapDir.h
Comment (yadij) [inline src/tests/TestSwapDir.h:44]:
```suggestion
#endif /* SQUID_SRC_TESTS_TESTSWAPDIR_H */
```

File: src/adaptation/ecap/Registry.h
Comment (yadij) [inline src/adaptation/ecap/Registry.h:12]:
```suggestion
#ifndef SQUID_SRC_ADAPTATION_ECAP_REGISTRY_H
#define SQUID_SRC_ADAPTATION_ECAP_REGISTRY_H

// TBD
```

or just drop this unused file entirely.

**Review by yadij — CHANGES_REQUESTED** [review]
Bunch of polish to the new additions.

**Comment by yadij** [comment]
> > > I had to hardcode ignoring headers in scripts/ , I couldn't think of any better way to avoid clobbering scripts/boilerplate.h without affecting its function
> > 
> > 
> > My [suggestion](https://github.com/squid-cache/squid/pull/1630#discussion_r1445233575) was to limit the corresponding fatal errors (and, hence, formatting) to headers in `src/`. I will evaluate your alternative approach when I review this PR.
> 
> we also have heaaders in include/ and lib/. In practice it seems to work well, the only file that shouldn't have header guards seems to be boilerplate.h. Looking forward to hear if you agree with my approach

I would like to see the other directories have at least a once-over by this script when the flag-day change goes in to make sure we are consistent everywhere.  `boilerplate.h` could be manually omitted at that time, with a followup rearrangement on how the boilerplate is done to better fit in with this script and with non-C/C++ files using it.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:11]: Already stale and unnecessary -- the interface is known/defined by this script being a source-maintenance plugin. And the script is not really meant to be used as a stand-alone program.

```suggestion
```

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:12]:
```suggestion
# source-maintenance plugin to standardize C/C++ header guard presence and spelling
```

or, at the very least:

```suggestion
# source-maintenance plugin to standardize the header guards
```

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:27]: Please declare this and other variables as late as possible. Doing so makes code more readable, clarifies its intent, reduces chances of leaving unused variables behind, etc.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:38]:
```suggestion
while (<>) {
```

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:37]:
```suggestion
# load preamble
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:11]: from its neighbour sort-includes.pl:

```
# USAGE: sort-includes.pl filename.cc >filename.cc.sorted
```

This said, I have zero attachment to a comment; removing it

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:12]: "standardise" is the proper spelling in UK English,
but sure

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:19]: removed

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:27]: This already is as late as possible; both variables are set in the loop
immediately below, and may not be scoped in that loop because they are
referenced outside the loop after it finishes

Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:27]: done (old stale comment had some skew in line numbers, sorry for that)

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:37]: changed

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:26]: done

Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:26]: changed

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:59]: reworked. Checks are now all consolidated in a single block

File: src/auth/digest/eDirectory/ldap_backend.h
Comment (kinkie) [reply src/auth/digest/eDirectory/ldap_backend.h:20]: done

File: src/auth/negotiate/kerberos/negotiate_kerberos.h
Comment (kinkie) [reply src/auth/negotiate/kerberos/negotiate_kerberos.h:154]: done

File: src/tests/STUB.h
Comment (kinkie) [reply src/tests/STUB.h:57]: done

File: src/tests/TestSwapDir.h
Comment (kinkie) [reply src/tests/TestSwapDir.h:44]: done

File: src/tests/TestSwapDir.h
Comment (kinkie) [reply src/tests/TestSwapDir.h:10]: done

File: src/auth/digest/LDAP/ldap_backend.h
Comment (kinkie) [reply src/auth/digest/LDAP/ldap_backend.h:21]: done by running the script

File: src/auth/basic/RADIUS/radius-util.h
Comment (kinkie) [reply src/auth/basic/RADIUS/radius-util.h:20]: done

File: src/auth/basic/NIS/nis_support.h
Comment (kinkie) [reply src/auth/basic/NIS/nis_support.h:14]: done

**Comment by yadij** [comment]
Applied the two polish bits myself. LGTM for the bits I have been reviewing. Now just waiting on @rousskov review.

**Review by yadij — APPROVED** [review]

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:34]: Please push first, before the `if` statements above. Then you will not have to remember to compensate with `+1` inside those `if` statements...

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:40]: The current PR code is more complex/tricky than it needs to be, but I do not insist on simplifying it further. In this particular case, the complexity may lead to overread of the accumulate array though:

```suggestion
    unless ($first_ifdef_pos < $#accumulate && $accumulate[$first_ifdef_pos+1] =~ /^#define\s+$current_guard/);
```

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:25]:
```suggestion
my $first_ifndef_pos = undef;
```

or, given the current (complex) code structure:

```suggestion
my $guard_ifdef_pos = undef;
```

... or something like that. I do not insist on this change, but I find the current "ifdef/ifndef" mismatch rather confusing.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:26]:
```suggestion
my $last_endif_pos = undef;
```

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:46]: I would replace everything except alphanumeric characters, but I do not insist on this change.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [inline scripts/maintenance/standard-header-guards.pl:47]: If we replace guard lines inside the accumulation array, we would not have to split array printing into several groups with complex start/end points. I do not insist on that approach though.

**Review by rousskov — APPROVED** [review]
Excellent progress. I think we are almost done here. Since I may not be able to re-review quickly, I am not formally requesting any changes, but please address as many of my change requests as possible before merging this.

File: scripts/maintenance/standard-header-guards.pl
Comment (rousskov) [reply scripts/maintenance/standard-header-guards.pl]: Yes, I am aware. Guess how many times that proper spelling variation is used in Squid sources?

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:25]: Good point. Replacing with $first_ifndef_pos

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:34]: fair point. doing that

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:40]: Overread can only happen if the first ifndef is the last line in the file (and if it is, source formatting is the last of our problems and this is likely ).

There are therefore only two cases where this can happen:
- no endif/truncated file - this is caught by the test in line 38
- last endif is before the first ifdef - I'm adding an explicit test for it
- I will add an explicit test for this scenario, with a more useful error message; I would expect it will never ever trigger tho

File: scripts/maintenance/standard-header-guards.pl
Comment (kinkie) [reply scripts/maintenance/standard-header-guards.pl:46]: For the files we currently have, I can't spot any obvious issues; I'll leave as-is for now

**Comment by kinkie** [comment]
> The current PR approach probably "works" but it is "neither here nor there" IMHO. I am not violently against it, but I think we can and should do better (by moving in either direction outlined in the bullets above).

Current approach only skips boilerplate.h

**Review by rousskov — COMMENTED** [review]
Thank you for addressing my primary concerns! Unfortunately, I do not have the time to check and simplify this further. I am not requesting any changes.

**Comment by kinkie** [comment]
In order to avoid disruption for followup PRs due to codeQL checks failing after merging this script,
I'll adopt this merge strategy:

1. I'll extract the header changes from this PR and land them as their own PR
2. I'll run the script once and land the produced changes as their own PR
3. I'll land the script

**Review by yadij — APPROVED** [review]

**Comment by kinkie** [comment]
Will readd backporting after having landed the v6-specific script-generated changes.

**Comment by squidadm** [comment]
queued for backport to v6

**Pull request closed without merging**
